File: solution3.py
# Time Complexity : O(N) where N is the number of words in s
# Space Complexity : O(M) where M is the number of unique words in s
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def word_pattern(self, pattern, s):
        index_dict = {}
        words = s.split()
        
        # edge case 
        if len(pattern) != len(words):
            return False
        
        for i in range(len(words)):
            c = pattern[i]
            w = words[i]

            # since string not hashable 
            char_key = 'char_{}'.format(c)
            char_word = 'word_{}'.format(w)
            
            if char_key not in index_dict:
                index_dict[char_key] = i
            
            if char_word not in index_dict:
                index_dict[char_word] = i 
        
            # return as soon as we find a mismatch
            if index_dict[char_key] != index_dict[char_word]:
                return False
        
        return True

File: solution3.java
package hashing;
//Time Complexity :  O(N) 
//Space Complexity :  O(M)
import java.util.HashMap;

public class solution3 {
	    public boolean wordPattern(String pattern, String s) {

	        HashMap<Character, String> charMap = new HashMap<>();
	        HashMap<String, Character> wordMap = new HashMap<>();
	        String[] words = s.split(" ");

	        if(words.length != pattern.length()) {
	            return false;
	        }

	        for (int i = 0; i < words.length; i++) {

	            char c = pattern.charAt(i);
	            String w = words[i];

	            if(!charMap.containsKey(c)){
	                if(wordMap.containsKey(w)) {
	                    return false;
	                }
	                else
	                {
	                    charMap.put(c, w);
	                    wordMap.put(w, c);
	                }
	            }
	            else
	            {
	                String mapped = charMap.get(c);
	                if (!mapped.equals(w)) {
	                    return false;
	                }
	            }
	        }
	        return true;

	    }
	}